src/main/java/org/jboss/modules/ModuleLoggerFinder.java

```diff
@@ -34,64 +43,172 @@
+    // Guarded by lock
+    private static LoggerFinder finder;
 
     @Override
     public System.Logger getLogger(final String name, final java.lang.Module module) {
-        return resolveFinder().getLogger(name, module);
+        if (activated) {
+            return finder.getLogger(name, module);
+        }
+        lock.lock();
+        try {
+            // Check if we're activated at this point, if so we need to return a logger from the resolved finder
+            if (activated) {
+                // This should be set at this point
+                return finder.getLogger(name, module);
+            }
+            return loggers.computeIfAbsent(name, s -> queueingFinder.getLogger(name, module));
+        } finally {
+            lock.unlock();
+        }
     }
 
     /**
-     * Sets the class loader to use when attempting to find the {@link LoggerFinder}.
+     * Activates the module logger and replaces any old delegating {@link System.Logger system loggers} with a logger
+     * from the new finder.
      *
      * @param cl the class loader to use
      */
-    static void setClassLoader(final ClassLoader cl) {
-        classLoader = cl;
-    }
-
-    private LoggerFinder resolveFinder() {
-        LoggerFinder result = finder;
-        if (result == null) {
-            synchronized (this) {
-                result = finder;
-                if (result == null) {
-                    try {
-                        final ClassLoader cl = classLoader;
-                        final ServiceLoader<LoggerFinder> loader = ServiceLoader.load(LoggerFinder.class, cl == null ? ClassLoader.getSystemClassLoader() : cl);
-                        for (LoggerFinder lf : loader) {
-                            if (!getClass().equals(lf.getClass())) {
-                                finder = result = lf;
-                                break;
-                            }
+    static void activate(final ClassLoader cl) {
+        // To keep order, we must observe the lock
+        lock.lock();
+        try {
+            if (!activated) {
+                try {
+                    final ServiceLoader<LoggerFinder> loader = ServiceLoader.load(LoggerFinder.class, cl == null ? ClassLoader.getSystemClassLoader() : cl);
+                    for (LoggerFinder lf : loader) {
+                        if (!ModuleLoggerFinder.class.equals(lf.getClass())) {
+                            finder = lf;
+                            break;
                         }
-                        if (result == null) {
-                            finder = result = defaultFinder.get();
+                    }
+                    if (finder == null) {
+                        finder = defaultFinder.get();
+                    }
+                } catch (Throwable ignore) {
+                    finder = defaultFinder.get();
+                }
+                // Process the queued loggers and drain the contents to new loggers from the finder found
```

Comment:
@jamezp Should there be a common record queue that gets drained? This will end up logging grouped by logger, not by order of record queuing.

Reply:
Yes, that is a good point. I'll fix that.

Reply:
Thanks.